oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -309,6 +336,37 @@ private static URL toURL(String url) {
 			}
 		}
 
+		private static class NoOpJwkSetCache implements JWKSetCache {
+			@Override
+			public void put(JWKSet jwkSet) {
+			}
+
+			@Override
+			public JWKSet get() {
+				return null;
+			}
+
+			@Override
+			public boolean requiresRefresh() {
+				return true;
+			}
+		}
+
+		private static class CachingResourceRetriever implements ResourceRetriever {
+			private final Cache cache;
+			private final ResourceRetriever resourceRetriever;
+
+			CachingResourceRetriever(Cache cache, ResourceRetriever resourceRetriever) {
+				this.cache = cache;
+				this.resourceRetriever = resourceRetriever;
+			}
+
+			@Override
+			public Resource retrieveResource(URL url) {
+				return cache.get(url.toString(), () -> resourceRetriever.retrieveResource(url));
```

Comment:
This might not work with some caches since `Resource` isn't `Serializable`. I'd recommend something closer to

```suggestion
				String jwks = cache.get(url.toString(), () -> resourceRetreiver.retreiveResource(url).getContent());
				return new Resource(jwks, "UTF-8");
```
return cache.get(url.toString(), () -> resourceRetriever.retrieveResource(url));

Reply:
Yes, I guess we'll have this issues with Redis cache for example, nice one, thanks